source/pkcs11.c

```diff
@@ -398,3 +426,312 @@ void aws_pkcs11_lib_release(struct aws_pkcs11_lib *pkcs11_lib) {
+
+except:
+finally:
+    aws_mem_release(pkcs11_lib->allocator, slot_id_array);
+    return success ? AWS_OP_SUCCESS : AWS_OP_ERR;
+}
+
+int aws_pkcs11_lib_open_session(
+    struct aws_pkcs11_lib *pkcs11_lib,
+    aws_pkcs11_t slot_id,
+    aws_pkcs11_t *out_session_handle) {
+
+    CK_SESSION_HANDLE session_handle;
+    CK_RV rv = pkcs11_lib->function_list->C_OpenSession(
+        slot_id, CKF_SERIAL_SESSION /*flags*/, NULL /*pApplication*/, NULL /*notify*/, &session_handle);
+    if (rv != CKR_OK) {
+        return s_raise_ck_error(pkcs11_lib, "C_OpenSession", rv);
+    }
+
+    /* success! */
+    AWS_LOGF_DEBUG(
+        AWS_LS_IO_PKCS11, "id=%p session=%lu: Session opened on slot %lu", (void *)pkcs11_lib, session_handle, slot_id);
+
+    *out_session_handle = session_handle;
+    return AWS_OP_SUCCESS;
+}
+
+void aws_pkcs11_lib_close_session(struct aws_pkcs11_lib *pkcs11_lib, aws_pkcs11_t session_handle) {
+    CK_RV rv = pkcs11_lib->function_list->C_CloseSession(session_handle);
+    if (rv == CKR_OK) {
+        AWS_LOGF_DEBUG(AWS_LS_IO_PKCS11, "id=%p session=%lu: Session closed", (void *)pkcs11_lib, session_handle);
+    } else {
+        /* Log the error, but we can't really do anything about it */
+        AWS_LOGF_WARN(
+            AWS_LS_IO_PKCS11,
+            "id=%p session=%lu: Ignoring C_CloseSession() failure. PKCS#11 error: %s (0x%08lX).",
+            (void *)pkcs11_lib,
+            session_handle,
+            s_ckr_str(rv),
+            rv);
+    }
+}
+
+int aws_pkcs11_lib_login_user(
+    struct aws_pkcs11_lib *pkcs11_lib,
+    aws_pkcs11_t session_handle,
+    const struct aws_string *optional_user_pin) {
+
+    CK_UTF8CHAR_PTR pin = NULL;
+    CK_ULONG pin_len = 0;
+    if (optional_user_pin) {
+        if (optional_user_pin->len > ULONG_MAX) {
+            AWS_LOGF_ERROR(AWS_LS_IO_PKCS11, "id=%p session=%lu: PIN is too long.", (void *)pkcs11_lib, session_handle);
+            return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        }
+        pin_len = (CK_ULONG)optional_user_pin->len;
+        pin = (CK_UTF8CHAR_PTR)optional_user_pin->bytes;
+    }
+
+    CK_RV rv = pkcs11_lib->function_list->C_Login(session_handle, CKU_USER, pin, pin_len);
+    if (rv != CKR_OK) {
+        return s_raise_ck_session_error(pkcs11_lib, "C_Login", session_handle, rv);
+    }
+
+    /* Success! */
+    AWS_LOGF_DEBUG(AWS_LS_IO_PKCS11, "id=%p session=%lu: User logged in", (void *)pkcs11_lib, session_handle);
+    return AWS_OP_SUCCESS;
+}
+
+/**
+ * Find the object that meets all criteria:
+ * - is private key
+ * - if match_label is non-null, then labels must match
+ * The function fails unless it finds exactly one object meeting all criteria.
+ */
+int aws_pkcs11_lib_find_private_key(
+    struct aws_pkcs11_lib *pkcs11_lib,
+    aws_pkcs11_t session_handle,
+    const struct aws_string *match_label,
+    aws_pkcs11_t *out_key_object_handle) {
+
+    bool success = false;
+
+    /* whether C_FindObjectsFinal() must be run */
```

Comment:
How do you decide this? Isnt it required always?

Reply:
added clarifying comments